src/gas/fetchFeeHistory.ts

```diff
@@ -0,0 +1,146 @@
+ * range requested.
+ */
+
+type FeeHistory = {
+  startBlockId: HexString;
+  blocks: BlockFeeHistory[];
+};
+
+/**
+ * @type BlockFeeHistory
+ *
+ * Historical data for gas and priority fees for a block.
+ * @property {Array<number>} baseFeePerGas - The base fee per gas for the block.
+ * @property {Array<number>} gasUsedRatio - A number between 0 and 1 that
+ * represents the ratio between the effective gas used and the gas limit for
+ * the block.
+ * @property {Map<number, number>} priorityFeesByPercentile - The priority
+ * fees paid for the transactions in the block that occurred at particular
+ * percentiles (using the gas used vs. gas limit ratio to determine those
+ * percentiles), then indexed by those percentiles.
+ */
+
+type BlockFeeHistory = {
+  baseFeePerGas: number;
+  gasUsedRatio: number;
+  priorityFeesByPercentile: Map<number, number>;
+};
+
+type EthFeeHistoryResponse = {
+  jsonrpc: string;
+  id: number;
+  result: {
+    oldestBlock: HexString;
+    baseFeePerGas: HexString[];
+    gasUsedRatio: number[];
+    reward: HexString[][];
+  };
+};
+
+/**
+ * Converts a hexadecimal string to a decimal number.
+ *
+ * @param hex - A string encoding a hexadecimal number (with a "0x") prefix.
+ * @returns The number in decimal.
+ */
+function hexToDec(hex: string): number {
+  return parseInt(new BN(stripHexPrefix(hex), 16).toString(10), 10);
+}
+
+/**
+ * Uses eth_feeHistory (an EIP-1559 feature) to obtain information about gas and priority fees from
+ * a range of blocks that appeared on a network, starting from a particular block and working
+ * backward in time. This can be used to gauge how congested the network is.
+ *
+ * @param args - The arguments.
+ * @param args.ethQuery - An EthQuery instance that wraps a provider for the network in question.
+ * @param args.endBlock - Which block to start with when forming the block range. Can be a block tag
+ * ("latest" for the latest successful block or "pending" for the latest pending block) or a block
+ * id.
+ * @param args.numberOfBlocks - How many blocks to jump back from the endBlock when forming the
+ * block range. Each array in the return value will be of this length.
+ * @param args.percentiles - A tuple of numbers between 1 and 100 which will advise how
+ * priorityFeesByPercentile in the return value will be formed. When Ethereum runs the
+ * eth_feeHistory method, for each block it is considering, it will calculate the ratio of gas used
+ * for each transaction compared to the total gas used for the block, then sort all transactions in
+ * the block by this ratio. For each percentile given, it will then find the transaction whose gas
+ * used ratio matches the percentile and capture that transaction's priority fee. Hence,
+ * priorityFeesByPercentile represents the priority fees of transactions at key gas used ratios.
+ * @returns The fee history data.
+ */
+export async function fetchFeeHistory({
```

Comment:
This function works different than the existing function in `util.ts` for fetching gas fee estimates in that instead of taking a URL it takes an EthQuery. We may change this to hit a Metaswap API instead of Infura, so this is subject to change. I just went with something for now.